### PR TITLE
Home work 8

### PR DESCRIPTION
**Pull request by UnitOmega13**

### CODE REVIEW

File: src/List.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+public interface List<T> {
+
+	void add(T value);
+	void add(T value, int index);
+	void addAll(List<T> list);
+	T get(int index);
+	void set(T value, int index);
+	T remove(int index); // возвращаем элемент, который удалили
```
Comment (Sofasmile) [inline src/List.java]: remove comments

File: src/NewLinkedList.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		return length == 0;
+	}
+
+	private void validIndex(int index) {
+		if (index < 0 || index > length - 1) {
+			throw new IndexOutOfBoundsException(index + "is not valid index");
+		}
+	}
+
+	private NewNode<T> getNode(int index) {
```
Comment (Sofasmile) [inline src/NewLinkedList.java:115]: add checking to index

File: src/NewLinkedList.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+		} else {
+			lastNewNode.setNext(newNode);
+			newNode.setPrevious(lastNewNode);
+			lastNewNode = newNode;
+		}
+		length++;
+	}
+
+	@Override
+	public void add(T value, int index) {
```
Comment (Sofasmile) [inline src/NewLinkedList.java:21]: if we write:
List<String> l = new NewLinkedList<>();
       l.add("1");
       l.add("2");
       l.add("3");
       l.add("2+", 2);
we will get:
1, 2, 2+, 3
test how work this method

File: src/NewLinkedList.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	}
+
+	@Override
+	public void addAll(List<T> list) {
+		for (int i = 0; i < list.size(); i++) {
+			add(list.get(i));
+		}
+	}
+
+	public void addAll(NewLinkedList<T> newLinkedList) {
```
Comment (Sofasmile) [inline src/NewLinkedList.java]: do we need this method here?

File: src/NewLinkedList.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+		lastNewNode.setNext(newLinkedList.firstNewNode);
+		newLinkedList.firstNewNode.setPrevious(this.lastNewNode);
+		length = length + newLinkedList.size();
+	}
+
+	@Override
+	public T get(int index) {
```
Comment (Sofasmile) [inline src/NewLinkedList.java:48]: the same about checking

File: src/NewLinkedList.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+	@Override
+	public T remove(T o) {
+		NewNode<T> newNode = firstNewNode;
+		if (o.equals(firstNewNode.value)) {
+			return removeFirst();
+		}
+		if (o.equals(lastNewNode.value)) {
+			return removeLast();
+		}
+		while (newNode != lastNewNode) {
```
Comment (Sofasmile) [inline src/NewLinkedList.java:87]: try use remove(int index) to simplify code

File: src/NewLinkedList.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+		NewNode<T> newNode = firstNewNode;
+		StringBuilder stringBuilder = new StringBuilder("NewLinkedLisr{");
+		for (int i = 0; i < length; i++) {
+			stringBuilder.append(newNode.value).append(" ");
+			newNode = newNode.next;
+		}
+		stringBuilder.append("}");
+		return stringBuilder.toString();
+	}
+}
```
Comment (Sofasmile) [inline src/NewLinkedList.java]: add a blank line

File: src/NewNode.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
+public class NewNode<T> {
```
Comment (Sofasmile) [inline src/NewNode.java]: better make it private static nested classes

Comment (UnitOmega13) [reply src/NewNode.java]: changed everything according to the comments

File: src/Aplication.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+public class Aplication {
+
+	public static void main(String[] args) {
+		List l = new NewLinkedList<>();
+		l.add("1");
+		l.add("2");
+		l.add("3");
+		l.add("2+", 2);
+		System.out.println(l);
```
Comment (SevostyanovGerman) [inline src/Aplication.java]: where testing other methods?

**Review by SevostyanovGerman — CHANGES_REQUESTED** [review]

File: src/Aplication.java (hunk 9 above)
Comment (UnitOmega13) [reply src/Aplication.java]: I added testing of other methods.

**Pull request closed without merging**